deepspeed/pt/zero_optimizer_stage1.py

```diff
@@ -598,28 +597,20 @@ def reduce_scatter_gradients(self,
 
                 all_comm_partitions.append(single_comm_all_partitions)
 
-            for p in my_params:
-                partitions = param_partition_map[p]
-                parts = []
-                for part in partitions:
-                    params, offsets = partition_param_map[part]
-                    found = False
-                    for p_idx, _p in enumerate(params):
-                        if p.__hash__() == _p.__hash__():
-                            found = True
-                            if offsets[p_idx][0] is not None:
-                                my_part = part.narrow(0,
-                                                      offsets[p_idx][0],
-                                                      offsets[p_idx][1])
-                                parts.append(my_part)
-                    assert found
-                if p is not None:
-                    updated_grad = _unflatten_dense_tensors(torch.cat(parts), [p])
-                    p.grad.copy_(updated_grad[0])
+            # stitch together all rank sub partitions for each comm idx
+            flat_comm_grads = []
+            for comm_idx, rank_partitions in enumerate(all_comm_partitions):
+                flat_comm_grads.append(torch.cat(rank_partitions))
+
+            flat_all_grads = torch.cat(flat_comm_grads)
+
+            # copy back reduced gradients but only those needed for this local rank
+            for param, updated_grad in zip(self.fp16_groups[i], _unflatten_dense_tensors(flat_all_grads, self.fp16_groups[i])):
```

Comment:
By flattening the gradients aren't we creating an extra copy of the gradients? At this point are we going to have 3 (or 4) copies of the gradients? i) in param.grad, ii) in all_comm_partitions, iii) flattented gradients, (iv) gradients unflattened from the flattened gradients if _unflatten_dense_tensors is creating a new copy. ) 

We should probably release all param.grad before flattening, then delete all_comm_partitions after flattening. That way we will have at most one copy (or two) of the gradients alive at any given time?